pulser/register.py

```diff
@@ -20,11 +20,13 @@
 import numpy as np
 from numpy.typing import ArrayLike
 from scipy.spatial import KDTree
-from typing import Any, Dict, Iterable, Optional, Mapping, cast
+from typing import Any, cast, Dict, Iterable, Optional, Mapping, Union
 
 import pulser
 from pulser.json.utils import obj_to_dict
 
+QubitId = Union[float, int, str]
```

Comment:
Instead of doing this, you should do:
```suggestion
QubitId = TypeVar('QubitId', float, int, str)
```

Change this on both files (or maybe define it here and import it in `sequence.py` from `pulser.register`) and see if it solves some of your issues. I will review the whole thing again after that

Reply:
It did not solve the problem and, in fact, it made things much worse. ☹️

I think using a simple alias is fine in this case : https://mypy.readthedocs.io/en/stable/kinds_of_types.html#type-aliases.

```
pulser\sequence.py:154: error: Type variable "pulser.sequence.QubitId" is unbound
pulser\sequence.py:154: note: (Hint: Use "Generic[QubitId]" or "Protocol[QubitId]" base class to bind "QubitId" inside a class)
pulser\sequence.py:154: note: (Hint: Use "QubitId" in function signature to bind "QubitId" inside a function)
pulser\sequence.py:158: error: Argument 1 to "set" has incompatible type "KeysView[QubitId]"; expected "Iterable[QubitId]"
pulser\sequence.py:160: error: Type variable "pulser.sequence.QubitId" is unbound
pulser\sequence.py:160: note: (Hint: Use "Generic[QubitId]" or "Protocol[QubitId]" base class to bind "QubitId" inside a class)
pulser\sequence.py:160: note: (Hint: Use "QubitId" in function signature to bind "QubitId" inside a function)
pulser\sequence.py:173: error: Incompatible return value type (got "Dict[QubitId, Union[Union[int, float, complex, str, bytes, generic], Sequence[Union[int, float, complex, str, bytes, generic]], Sequence[Sequence[Any]], _SupportsArray]]", expected "Dict[float, Union[Union[int, float, complex, str, bytes, generic], Sequence[Union[int, float, complex, str, bytes, generic]], Sequence[Sequence[Any]], _SupportsArray]]")
pulser\sequence.py:173: error: Incompatible return value type (got "Dict[QubitId, Union[Union[int, float, complex, str, bytes, generic], Sequence[Union[int, float, complex, str, bytes, generic]], Sequence[Sequence[Any]], _SupportsArray]]", expected "Dict[int, Union[Union[int, float, complex, str, bytes, generic], Sequence[Union[int, float, complex, str, bytes, generic]], Sequence[Sequence[Any]], _SupportsArray]]")
pulser\sequence.py:173: error: Incompatible return value type (got "Dict[QubitId, Union[Union[int, float, complex, str, bytes, generic], Sequence[Union[int, float, complex, str, bytes, generic]], Sequence[Sequence[Any]], _SupportsArray]]", expected "Dict[str, Union[Union[int, float, complex, str, bytes, generic], Sequence[Union[int, float, complex, str, bytes, generic]], Sequence[Sequence[Any]], _SupportsArray]]")
pulser\sequence.py:304: error: Item "float" of "Union[Set[float], float]" has no attribute "__iter__" (not iterable)
pulser\sequence.py:304: error: Item "int" of "Union[Set[int], int]" has no attribute "__iter__" (not iterable)
```

Reply:
The problem with using just an alias for an `Union` is that it's inappropriate to be used as an argument to container types like `Dict` and `Set`, for example. It's only fine as an alias for the QubitId on its own

Reply:
But I believe you're right in saying that `TypeVar` is not appropriate either, now that I'm digging deeper into the docs (I'm sorry, I'm new to all this stuff too). I thought it was a solution to the problem I was encountering before of using `Union` to specify multiple possible argument types for `Dict`

Reply:
I found a way to put it that passed all the tests... I'll push it now. I've been using `QubitId = Union[int, str]` and it seems to work fine with `dict`.

Reply:
Alright, let's go with it then, perhaps the error I encountered was particular to the case I was dealing with.